### PR TITLE
Registrar NFT auth, OwnerAction signature binding, and minted==1 boundary fixes

### LINKED ISSUE #17: Replace registrar HNS-signature auth with a token/NFT check in tld_registrar
## Context

**Current**: `registrar_hns_key` is a validator parameter on
`tld_registrar`, checked via `verify_tld_signature` in `mint
RegisterTLD` and `spend RegistrarAction`. Baked into the deployed address
permanently — no rotation or recovery without a full redeploy and
migration of every registered TLD.

**Should be**: registrar authority proven by spending a designated
token/NFT as a tx input instead of a signature. Rotating the registrar
becomes "send the token to a new wallet" — no redeploy, no migration.

**Future option**: the token can later be locked at its own validator
requiring an HNS signature to release it, re-introducing signature-gated
control at the custody layer without touching `tld_registrar` again.

Scoped to **registrar** auth only. TLD *owner* auth (`owner_hns_key`/
`owner_signature`) is per-TLD datum state, not a validator parameter, and
is unaffected.

## Desired behavior

`registrar_hns_key` → `registrar_token_policy_id`/`tn` as validator
params; `verify_tld_signature(...)` → `token_in_inputs(...)` at both
check sites. Needs a one-shot minting policy for the token, and the
registration script/tooling updated to spend the token instead of
signing.

## Acceptance criteria

- [ ] Validator params + both check sites updated; unaffected paths
      (`RegistrarAction` mint/burn, `OwnerAction`) verified unchanged
- [ ] One-shot registrar-token minting policy added
- [ ] `aiken check` passes with tests covering token present/absent on
      both registration and deregistration
- [ ] Doc note added/updated on the new registrar-auth model

### PR DESCRIPTION
**Pull request by adrian1-dot**
## Summary

- Registrar authorization moved from an HNS signature check to a bearer NFT check (`registrar_token`), dropping the `registrar_signature` fields from `RegisterTLD`/`RegistrarAction`.
- `OwnerAction`'s Handshake signature is now bound to the token destination address (`receiver_address`) and the spend's own `output_reference`, closing a replay/redirect gap where the old signature (over `tld` alone) could be reused against a different destination or a later transaction.
- `InitRemoveReference` and `BurnReference` now correctly bracket `minted == 1` between them — previously a gap allowed the last reference to be burned early (orphaning the registration permanently) or the merge path to fire one step too early.
- Updated `hns-sig/sign.js` and `scripts/02-register-tld.sh`/`03-mint-tld.sh` to match the new redeemer shapes and signed-message format. Updated the architecture docs accordingly.

Fixes #17, #18, #19, #20.

## Implementation note on #17

#17 asks for the registrar NFT check to be performed on tx *inputs*. This implementation checks *outputs* instead (`policy_id_present_in_outputs`) — decided as the sounder and cheaper option: the NFT still has to be consumed and recreated somewhere in the tx for the check to pass, so the guarantee is equivalent, without requiring the registrar to additionally prove which UTxO it came from.

## Test plan

- `aiken check`: 32/32 passing, 0 failing.
- New coverage: `fail_tld_burn_missing_registrar_token` (registrar NFT absent on deregistration), `fail_tld_reference_burn_with_minted_one` (merge guard boundary), `fail_tld_reference_init_remove_burn_with_minted_two` (final-burn guard boundary), plus `OwnerAction` destination/replay tests (`fail_owner_action_wrong_destination`, `fail_owner_action_missing_user_token_output`, `fail_owner_action_replayed_signature_new_outref`).

---
## Summary by cubic
Replaces registrar authorization from a Handshake signature with a bearer registrar NFT and binds `OwnerAction` signatures to the destination and the spent registration UTxO. This blocks replay/redirect attacks and fixes off-by-one errors at the merge and final-burn boundaries.

- Adds a `registrar_token` one-shot NFT under a new `registrar_nft` mint policy (anchored to `output_reference`); `tld_registrar` now checks for the registrar policy in transaction outputs on deregistration and drops the registrar secp256k1 signature check.
- Redeemers: `RegisterTLD` removes `registrar_signature`; `RegistrarAction` is empty; `OwnerAction` adds `receiver_address` and signs blake2b_256(tld ++ serialiseData(receiver_address) ++ serialiseData(output_reference)); enforces the minted user token lands at `receiver_address`.
- Corrects guards: merges require minted > 1; the final burn only succeeds on a 1 → 0 transition. Adds negative tests for missing registrar NFT, minted==1 merge, and signature replay/redirect; keeps a regression proving `tld_reference` requires +1 when minted==0.
- Tooling/docs: updates signing (`hns-sig/sign.js` now signs the new message and accepts optional CBOR args), scripts (`02-register-tld.sh`/`03-mint-tld.sh`), and architecture docs (outputs-based registrar NFT check; distinguish registrar NFT vs registration token; include `sld_reference.ak`). Fixes the redeemer-mapping table (removes `registrar_signature`, adds `receiver_address` to `OwnerAction`), the `verify_tld_signature` snippet, and the filename to `registrar_nft.ak`; removes a duplicate test. Upgrades `aiken-lang/stdlib` to `v3.1.0` and the compiler to `v1.1.23`.

**Migration**
- Mint and carry the `registrar_token` through registration; ensure it appears in deregistration outputs.
- Regenerate `OwnerAction` signatures per transaction using the exact `receiver_address` and `output_reference`, and ensure the minted user token is output to `receiver_address`.

<sup>Written for commit 2446847791be0ad7ec02ef83ba204a12f72fbd05. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/blinklabs-io/dns-contracts/pull/21?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Registration and deregistration now use a bearer registrar NFT for authorization.
  * Initial activation requires a one-time owner signature tied to the destination address and registration reference.
  * Subsequent actions use bearer user tokens, with stronger protection against signature replay or incorrect destinations.
  * TLD reference-token removal now supports merges when multiple reference outputs exist.

* **Documentation**
  * Updated architecture, validation, and registration guidance to reflect the revised authorization and activation flows.

* **Bug Fixes**
  * Improved validation for missing authorization tokens, invalid destinations, and incomplete registration transactions.

### CODE REVIEW

File: hns-sig/sign.js [hunk 1]
```diff
@@ -13,9 +13,17 @@ do {
...
+const tldHex = Buffer.from('hello-handshake', 'utf8').toString('hex');
+const receiverAddressCborHex =
+  'd8799fd8799f581ccf2020680b6315ff98ffdddde4400839a628e2360a1d1a20ed519439ffd87a80ff';
+const outputReferenceCborHex =
+  'd8799f58200fd923ca5e7218c4ba3c3801c26a617ecdbfdaebb9c76ce2eca166e7855efbb800ff';
+
+const message = Buffer.from(
+  tldHex + receiverAddressCborHex + outputReferenceCborHex,
+  'hex'
+);
```
[the remark this answers is not in the document]
Comment (adrian1-dot) [reply hns-sig/sign.js:31]: Applied with a twist: made `receiverAddressCborHex`/`outputReferenceCborHex` optional CLI args (`node sign.js <addr> <outref>`) that default to the existing test fixture values, rather than requiring them unconditionally — keeps the tool usable standalone while allowing real values for an actual transaction.

File: onchain/validators/tld_registration/tld_registrar.ak [hunk 2]
```diff
@@ -34,62 +37,86 @@ validator tld_registrar(
...
           )
+        } else {
+          True
+        }
+
+        // Must mint or burn - zero changes are prohibited to prevent no-op transactions
+        expect !(tld_reference_mint_am == 0)
+        let registration_datum = InlineDatum(
+          TLDRegisterDatum(tld, owner_hns_key, minted + tld_reference_mint_am),
+        )
```
[the remark this answers is not in the document]
Comment (adrian1-dot) [reply onchain/validators/tld_registration/tld_registrar.ak:112]: Checked this carefully rather than adding the guard outright. Reachability: when `minted == 0`, `tld_reference.ak`'s own mint handler (InitRemoveReference) independently requires `check_init_minting(mint, policy_id, tld, tld_r_tn, 1)` — i.e. exactly +1 — and the ledger requires that handler to validate whenever `tld_reference_cs` tokens are minted/burned in the same tx, since both validators read the same input datum for `minted`. So a negative `tld_reference_mint_am` at `minted == 0` can never appear in a transaction that actually goes through.

Added two tests to prove this rather than assert it: `success_owner_action_alone_allows_negative_minted` shows OwnerAction's own checks really don't bound the sign (confirming the observation is correct in isolation), and `fail_owner_action_negative_minted_rejected_by_tld_reference` shows the same transaction is rejected once tld_reference's mint handler is required to also pass. Left the guard itself out since it would be dead code for an already-unreachable state, but happy to add it if you'd rather have the defense-in-depth.

File: onchain/lib/types.ak [hunk 3]
```diff
@@ -13,11 +14,10 @@ pub type RegistrarRedeemer {
   RegisterTLD {
     tld: ByteArray,
     owner: ByteArray,
-    registrar_signature: ByteArray,
     tld_reference_policy_id: PolicyId,
   }
-  RegistrarAction { registrar_signature: ByteArray }
-  OwnerAction { owner_signature: ByteArray }
+  RegistrarAction
```
[the remark this answers is not in the document]
Comment (adrian1-dot) [reply onchain/lib/types.ak:19]: Verified this against actual behavior: `aiken check` compiles and runs 33/33 tests clean with `RegistrarAction { .. } ->` in place right now. This particular record-pattern syntax on a nullary constructor is apparently accepted by Aiken, not a compile error — false positive.

File: scripts/02-register-tld.sh [hunk 4]
```diff
@@ -22,15 +22,11 @@ SLD_REFERENCE_REF_TX="ef635b55fce6abc39cd4c843722d9d574cb719114e224f2cd1c8747d5a
 OWNER_HNS_VKEY=$(cat $WALLET_PATH/owner1.hns | jq -r '.publicKey')
 echo "OWNER_HNS_VKEY: $OWNER_HNS_VKEY"
 
-USER_HNS_SIG=$(cat $WALLET_PATH/$USER.hns | jq -r '.signature')
-echo "USER_HNS_SIG: $USER_HNS_SIG"
-
-mint_red=$(jq -n --arg ref_cs "$CS_TLD" --arg vkey $OWNER_HNS_VKEY --arg sig $USER_HNS_SIG '{
+mint_red=$(jq -n --arg vkey "$OWNER_HNS_VKEY" --arg ref_cs "$CS_TLD" '{
```
[the remark this answers is not in the document]
Comment (adrian1-dot) [reply scripts/02-register-tld.sh:25]: Confirmed accurate

File: hns-sig/sign.js [hunk 5]
```diff
@@ -13,9 +13,22 @@ do {
...
 
-// Message and hash
-const message = Buffer.from('hello-handshake', 'utf8');
-//const hash = SHA256.digest(message);
+// message = tld ++ serialiseData(receiver_address) ++ serialiseData(output_reference)
+// receiver_address/output_reference CBOR default to the Aiken test fixtures
+// (mock_pub_key_address("u"), mock_utxo_ref("0", 0)) but can be overridden:
+//   node sign.js <receiverAddressCborHex> <outputReferenceCborHex>
+const tldHex = Buffer.from('hello-handshake', 'utf8').toString('hex');
+const receiverAddressCborHex =
```
[the remark this answers is not in the document]
Comment (adrian1-dot) [reply hns-sig/sign.js:21]: Confirmed accurate — the hardcoded default is the Aiken test fixture's address (stake=None by design), never meant to equal 03-mint-tld.sh's real stake-bearing address. That's exactly why the script's own comment says the signature has to be regenerated per real UTxO, and why the CLI-arg override exists. sign.js itself is demo-only and won't land in production use, so leaving as-is.

[the remark this answers is not in the document]
Comment (adrian1-dot) [reply hns-sig/sign.js:21]: Confirmed accurate and the most substantive finding here — the tool has no way to sign with a real owner key at all. Deferred: sign.js is demo-only and won't land in production use.

File: hns-sig/sign.js [hunk 6]
```diff
@@ -13,9 +13,22 @@ do {
...
+  process.argv[2] ||
```
[the remark this answers is not in the document]
Comment (adrian1-dot) [reply hns-sig/sign.js:22]: Confirmed accurate — hex arg validation missing, could silently produce a signature that never verifies. Deferred along with the rest of sign.js: it's demo-only and won't land in production use.

File: onchain/validators/tld_registration/tests/tld_reference.ak [hunk 7]
```diff
@@ -130,7 +130,24 @@ fn tld_reference_remove_additional_succeed_tx() {
   }
 }
 
-test success_tld_reference_spend_remove_update() {
+fn tld_reference_remove_additional_fail_tx() {
```
[the remark this answers is not in the document]
Comment (adrian1-dot) [reply onchain/validators/tld_registration/tests/tld_reference.ak]: Confirmed and fixed — renamed the helper to `..._minted_two_tx` since it's used by a success test, and while cleaning this up also found the paired `_succeed_tx` helper/test was a byte-for-byte duplicate of a test added earlier in this same PR (see the other cubic thread on this file) — removed it rather than just renaming.

File: onchain/validators/tld_registration/tests/tld_reference.ak [hunk 8]
```diff
@@ -130,7 +130,24 @@ fn tld_reference_remove_additional_succeed_tx() {
...
+      Input(mock_utxo_ref("0", 1), create_tld_reference_output([], ["c"], "")),
+      Input(mock_utxo_ref("12", 12), create_tld_user_token_output),
+      Input(mock_utxo_ref("0", 2), create_tld_registrar_output("", 2)),
+    ],
+    outputs: [create_tld_reference_output([], ["a", "b", "c"], "")],
+    mint: from_asset(tld_reference_cs, create_reference_token_tn(tld), -1),
+  }
+}
+
+test fail_tld_reference_spend_remove_update() fail {
```
[the remark this answers is not in the document]
Comment (adrian1-dot) [reply onchain/validators/tld_registration/tests/tld_reference.ak]: Confirmed and fixed — removed `fail_tld_reference_spend_remove_update` and its dedicated helper, keeping the more descriptively named `fail_tld_reference_burn_with_minted_one`.

File: docs/architecture/smart-contract-architecture.md [hunk 9]
```diff
@@ -771,11 +790,15 @@ These are precise observations about the code as written - useful when
...
-  The merge tests exercise it with a registration datum whose `minted == 1`
-  ([`tests/tld_reference.ak:116`](../../onchain/validators/tld_registration/tests/tld_reference.ak#L116)).
+- `BurnReference` (merge) guard is `minted > 1`
+  ([`tld_reference.ak:278`](../../onchain/validators/tld_registration/tld_reference.ak#L278)),
+  not literal equality with `1`. This is the semantically necessary condition
+  for the merge path: `BurnReference` requires two reference-token inputs
+  (`expect [Input, Input] = ...`), which is only possible when `minted >= 2`;
+  a guard of literal `minted == 1` would make the branch permanently
+  unreachable. Exercised by `success_tld_reference_burn_with_minted_two`
+  ([`tests/tld_reference.ak:188`](../../onchain/validators/tld_registration/tests/tld_reference.ak#L188))
```
[the remark this answers is not in the document]
Comment (adrian1-dot) [reply docs/architecture/smart-contract-architecture.md]: Confirmed and fixed — updated the line reference (it drifted twice more from subsequent edits, now points at the correct line).

File: docs/architecture/smart-contract-architecture.md [hunk 10]
```diff
@@ -361,13 +369,18 @@ flowchart LR
...
+TLD concatenated with the `serialiseData` CBOR of the intended token
+destination and this spend's own `output_reference`
+([`utils.ak:23`](../../onchain/lib/utils.ak#L23),
+[`tld_registrar.ak`](../../onchain/validators/tld_registration/tld_registrar.ak)),
+so a captured signature can't be redirected to a different address or
+replayed against a later transaction:
 
 ```aiken
-pub fn verify_tld_signature(verification_key, tld, signature) -> Bool {
+pub fn verify_tld_signature(verification_key, message, signature) -> Bool {
```
[the remark this answers is not in the document]
Comment (adrian1-dot) [reply docs/architecture/smart-contract-architecture.md]: Confirmed and fixed — the doc snippet now matches the real function signature (still takes a plain `tld: ByteArray` and hashes it directly), with a note that the message composition happens at the tld_registrar.ak call site.

File: docs/architecture/validation-method.md [hunk 11]
```diff
@@ -108,12 +108,14 @@ Our validation system enables Handshake domain owners to:
 
 ## Technical Implementation Details
 
-The validation mechanism is implemented across two validators:
+The validation mechanism is implemented across four on-chain pieces:
 
-**tld_registrar.ak**: Handles initial registration and signature verification
+**registrar_token.ak**: Mints the registrar auth NFT used by the registrar wallet
```
[the remark this answers is not in the document]
Comment (adrian1-dot) [reply docs/architecture/validation-method.md]: Confirmed and fixed — relabeled to `registrar_nft.ak` (validator `registrar_token`).

File: docs/architecture/smart-contract-architecture.md [hunk 12]
```diff
@@ -61,7 +61,6 @@ The core idea is "verify once, then tokenize":
 flowchart TB
```
[the remark this answers is not in the document]
Comment (adrian1-dot) [reply docs/architecture/smart-contract-architecture.md:61]: Confirmed and fixed — updated the redeemer table (dropped `registrar_signature` from both rows, added `receiver_address` to OwnerAction) and the section 15 bullet.
